### DISCUSSION TITLE
Is it possible to test a Github action without publishing it on marketplace?

### DISCUSSION THREAD

**Discussion by aslisabanci** [community/community#26073] [Actions]
<p>My current develop-test iterations while creating a Github action is as follows:</p>
<ol>
<li>Develop the github action, create a release and publish it on marketplace, as v0.1.0 for instance.</li>
<li>Have the sample test repository to use this action’s v0.1.0 and test the workflow</li>
<li>If the action doesn’t behave as desired, push new updates on its repository, create another release and publish another version on marketplace as v0.1.1 for instance.</li>
<li>Update the test repo’s workflow file to use v0.1.1 and test the workflow again</li>
<li>Repeat and complain about how ineffective this is :slight_smile:
</li>
</ol>
<p>As you can imagine, step <span class="hashtag">#3</span> happens a lot because this is a development cycle and you may not know how to do some operations, how to pass variables, how to run steps/jobs, etc… all in advance. So you iterate tons of times when you’re creating a Github action.</p>
<p>Is there a way to shorten this lifecycle? Can we somehow test the actions we create, before publishing them on the marketplace and having to create version after version? This would immensely reduce the development time for your action creators.</p>
<p>Thank you in advance!</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/aslisabanci">@aslisabanci</a>,</p>
<p>Based on the <a href="https://docs.github.com/en/actions/creating-actions/publishing-actions-in-github-marketplace#about-publishing-actions" rel="nofollow noopener">doc</a> of publishing marketplace, your action repository should be public.</p>
<p>Github actions supports git ref/sha/tag version for the usage: {owner}/{repo}@{ref}(doc <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsuses" rel="nofollow noopener">here</a>). Hence, before you publich the action to marketplace, you can use <code>{owner}/{repo}@sha</code> to have a test. Publish it untill it works as expected.</p>
<p>Thanks</p>

**Reply by thejeff77** [reply]
Could you give an example? The `{owner}/{repo}@sha` doesn't give the action name in your example.

For example, the only thing I can get working is a fully qualified path: `{org}/{repo}/.github/actions/my-action@sha` Is this supposed to work when your custom action is in /github/actions? -- `{org}/{repo}/my-action@sha`?

**Comment by aslisabanci** [comment]
<p>Thanks for the answer <a class="mention" href="/weide-zhou">@weide-zhou</a>!</p>
<p>Yes my action repository is public, because I cannot otherwise publish it on the marketplace as you said. I didn’t know that {owner}/{repo}@{ref} usage works <strong>before/without</strong> publishing the action on the marketplace though. If this is made clear on the <a href="https://docs.github.com/en/actions/creating-actions" rel="nofollow noopener">“Creating Actions” section of the documentation</a>, I’m sure it would help many others suffering from the same thing as me.</p>

**Comment by weide-zhou** [comment]
<p>Thanks for your quick reply <a class="mention" href="/aslisabanci">@aslisabanci</a>!</p>
<p>According to the <a href="https://docs.github.com/en/actions/getting-started-with-github-actions/about-github-actions#contacting-support" rel="nofollow noopener">policy</a>, it’s recommended to raise a feedback ticket in below link where github product manager will take a review, it will promote the improvement of the doc.</p>
<p><a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" class="onebox" target="_blank" rel="nofollow noopener">https://support.github.com/contact/feedback?contact[category]=actions</a></p>
<p>You can also mark the answer then it could help other guys who have same query. :grinning:</p>

**Comment by Simran-B** [comment]
<p>Private actions are on the roadmap for Q1 2021:<br>
</p><aside class="onebox githubissue">
  <header class="source">
      <a href="https://github.com/github/roadmap/issues/74" target="_blank" rel="nofollow noopener">github.com/github/roadmap</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/github/roadmap/issues/74" target="_blank" rel="nofollow noopener">Actions: Use actions from private repositories</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2020-07-24" data-time="17:15:35" data-timezone="UTC">05:15PM - 24 Jul 20 UTC</span>
      </div>


      <div class="user">
        <a href="https://github.com/github-product-roadmap" target="_blank" rel="nofollow noopener">
          <img alt="github-product-roadmap" src="https://user-images.githubusercontent.com/7819991/181095087-666d1171-72d7-4c8a-b801-bf344c97daf9.png" class="onebox-avatar-inline" width="20" height="20">
          github-product-roadmap
        </a>
      </div>
    </div>
  </div>
</div>

<div class="github-row">
  <p class="github-content">Summary
This feature enables users to reference actions in private repositories, without having to publish those actions to the GitHub Marketplace.
Intended Outcome
Currently,...</p>
</div>

<div class="labels">
    <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">actions</span>
    <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">all</span>
    <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">cloud</span>
    <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">code-to-cloud</span>
    <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">ga</span>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
